ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,370 @@
+     */
+    function initializeValidatorSet(
+        ValidatorMessages.SubnetConversionData calldata subnetConversionData,
+        uint32 messsageIndex
+    ) external;
+
+    /**
+     * @notice Initiate a validator registration by issuing a RegisterSubnetValidatorTx Warp message
+     * @param nodeID The ID of the node to add to the Subnet
+     * @param weight The weight of the node on the Subnet
+     * @param expiry The time after which this message is invalid
+     * @param signature The Ed25519 signature of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[timestamp]
+     */
+    function initiateValidatorRegistration(
+        bytes32 nodeID,
+        uint64 weight,
+        uint64 expiry,
+        bytes memory signature
+    ) external returns (bytes32);
+
+    /**
+     * @notice Resubmits a validator registration message to be sent to P-Chain.
+     * Only necessary if the original message can't be delivered due to validator churn.
+     * @param validationID The validationID attached to the registration message
+     */
+    function resendValidatorRegistrationMessage(bytes32 validationID) external returns (bytes32);
+
+    /**
+     * @notice Completes the validator registration process by returning an acknowledgement of the registration of a
+     * validationID from the P-Chain.
+     * @param messageIndex The index of the Warp message to be received providing the acknowledgement.
+     */
+    function completeValidatorRegistration(uint32 messageIndex) external;
+
+    /**
+     * @notice Initiate a validator weight update by issuing a SetSubnetValidatorWeightTx Warp message.
+     * If the weight is 0, this initiates the removal of the validator from the Subnet. An uptime proof can be
+     * included. This proof might be required to claim validator rewards (handled by the security module).
+     * @param nodeID The ID of the node to modify
+     * @param weight The new weight of the node on the Subnet
+     * @param includesUptimeProof Whether the uptime proof is included in the message
+     * @param messageIndex The index of the Warp message containing the uptime proof
+     */
+    function initiateValidatorWeightUpdate(
+        bytes32 nodeID,
+        uint64 weight,
+        bool includesUptimeProof,
+        uint32 messageIndex
+    ) external;
+
+    /**
+     * @notice Completes the validator weight update process by returning an acknowledgement of the weight update of a
+     * validationID from the P-Chain.
+     * @param messageIndex The index of the Warp message to be received providing the acknowledgement.
+     */
+    function completeValidatorWeightUpdate(uint32 messageIndex) external;
+}
+```
+
+#### About `Validation`s
+
+A `Validation` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator weight changes during the same validation.
```

Comment:
Tracking independent periods each time the validator's weight changes during the same validation is interesting.

I find it more intuitive to consider a validator's own (initial) weight as static for the entire validation period, and consider any changes to the weight as delegations. I think that makes it easier for the security module (or whichever contract would be responsible for potential rewards) to compute the rewards for a given validator or delegator.

Reply:
It is indeed more complicated.

What I had in mind was the auto-compounding workflow for validators. Ending a validation will most likely trigger a delay period during which tokens will be idle. Offering the capacity to change the weight during a validation allows us to circumvent this problem.

In my implementation, I compute the average of the validator weight during the whole validation period in `_notifySecurityModuleValidatorWeightUpdate`. Happy to have your thoughts: https://github.com/suzaku-network/suzaku-contracts-library/blob/1d13efa366ddd4b288a884351402c3492f65e15d/contracts/src/contracts/ACP99/ACP99Manager.sol#L395